### ISSUE TITLE
Vitepress' router.js prevents event trackers accurately recording 

### ISSUE THREAD

**Issue by joepavitt** [vuejs/vitepress#3028]
### Describe the bug

I first noticed this working with [PostHog](https://posthog.com/) on the [documentation site for Node-RED Dashboard 2.0](https://dashboard.flowfuse.com/).

They have an `autocapture` feature which adds adds a `click` handler to all DOM elements ([source](https://github.com/PostHog/posthog-js/blob/a67ede333314ad5263ac97446480d6d1c22c072b/src/autocapture.ts#L186)) in order to track clicks of links, etc. in an application/website. 

The problem I noticed was that any link clicks were not registering to the current page, but instead, the target page, i.e. if I clicked our "Getting Started" button on our home page, it was recording that the "Getting Started" button was clicked at `/getting-started` (which doesn't exist there).

Upon further investigation, I found, looking at the Event Listeners of a given link on our website, that there is a Vitepress `router.js` function, which updates the URL _before_ any of the `document` handlers (including PostHog, Google Analytics, etc) can run their code, meaning that the events in inaccurately recorded. 

<img width="342" alt="Screenshot 2023-09-28 at 16 56 08" src="https://github.com/vuejs/vitepress/assets/99246719/368e7b8e-3be6-4d71-81b7-885775058a3d">

I'm not entirely sure how to fix this, but we use PostHog on our website, with no issues, and I also use PostHog in a Vue App (our FlowFuse platform) with `vue-router`, and there are no issues there either, it is _just_ when using Vitepress due to the `window.addEventListener('click', (e) => { })` , hence me raising the issue here.

I suspect (looking at the Vitepress code) this may even be more niché in that it's also only applicable when using `.html` extensions in the URLs, but we host on GitHub pages, where I understand this is the only option?

I can also confirm that removing the event handler completely does then allow PH to record data correctly, albeit, it then completely refreshes the page, rather than smooth SPA-style page transitions.

### Reproduction

Add posthog to any Vitepress site, by including:

```js
[
        'script',
        {},
        `
          if (!window.posthog){!function(t,e){var o,n,p,r;e.__SV||(window.posthog=e,e._i=[],e.init=function(i,s,a){function g(t,e){var o=e.split(".");2==o.length&&(t=t[o[0]],e=o[1]),t[e]=function(){t.push([e].concat(Array.prototype.slice.call(arguments,0)))}}(p=t.createElement("script")).type="text/javascript",p.async=!0,p.src=s.api_host+"/static/array.js",(r=t.getElementsByTagName("script")[0]).parentNode.insertBefore(p,r);var u=e;for(void 0!==a?u=e[a]=[]:a="posthog",u.people=u.people||[],u.toString=function(t){var e="posthog";return"posthog"!==a&&(e+="."+a),t||(e+=" (stub)"),e},u.people.toString=function(){return u.toString(1)+".people (stub)"},o="capture identify alias people.set people.set_once set_config register register_once unregister opt_out_capturing has_opted_out_capturing opt_in_capturing reset isFeatureEnabled onFeatureFlags getFeatureFlag getFeatureFlagPayload reloadFeatureFlags group updateEarlyAccessFeatureEnrollment getEarlyAccessFeatures getActiveMatchingSurveys getSurveys".split(" "),n=0;n<o.length;n++)g(u,o[n]);e._i.push([i,s,a])},e.__SV=1)}(document,window.posthog||[]);
          posthog.init('${process.env.VITE_POSTHOG_APIKEY}',{api_host:'https://eu.posthog.com', capture_pageview: false, capture_pageleave: false})}
        `
      ],
```

in the `head` configuration in `.vitepress/config.js`, and replace the APIKEY accordingly.

Then clicking around the site, you can see in PostHog's Event Explorer, that the events are mis-recorded. You can also run `posthog.debug()` in the browser's console, and click around to see the recorded events, and relevant properties (`properties.$current_url`) recorded.

### Expected behavior

URL shouldn't be updated _before_ the page has actually transitioned, and before any other JS click handlers for that page have run.

### System Info

```Text
System:
    OS: macOS 13.0
    CPU: (10) arm64 Apple M1 Pro
    Memory: 124.02 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    Yarn: 1.22.15 - /usr/local/bin/yarn
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
    pnpm: 8.1.1 - ~/.nvm/versions/node/v18.15.0/bin/pnpm
Browsers:
    Chrome: 117.0.5938.92
    Safari: 16.1
npmPackages:
    vitepress: ^1.0.0-beta.7 => 1.0.0-rc.12
```

### Validations

- [X] Check if you're on the [latest VitePress version](https://github.com/vuejs/vitepress/releases/latest).
- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

**Comment by brc-dd** [comment]
I don't think appearing at bottom in that list means it will be executed "after" all others. In fact things near the bottom are executed first because of hierarchy (click bubbles up, event listener on document should be executed before that on window).

**Comment by brc-dd** [comment]
can you try using a patch version of vitepress with `{ capture: true }` removed in `node_modules/vitepress/dist/client/app/router.js` (use pnpm patch or patch-package).

**Comment by joepavitt** [comment]
Hi @brc-dd - I can confirm that does then record the correct `current_url` of the `autocapture` events, however, it no longer runs the `vue-router`, so am getting full page refreshes, with full page flashing, rather than the nice SPA transitions.

Also now get no `pageleave` or `pageview` events. Although I can turn those back on with the default PostHog behaviour which should capture them nicely again, rather than watching the `router` as I am now, and manually sending them. The removal of that code seems to stop the app running like an SPA in navigation?

**Comment by brc-dd** [comment]
Yeah, I'm not sure what would be the fix then. We don't use vue-router.  Vue-router is behaving differently because that needs you to use its own link component rather than normal anchor tag. And it's likely adding an event listener for each link, instead of delegating them.

**Comment by joepavitt** [comment]
> it's likely adding an event listener for each link, instead of delegating them.

This seems to be exactly the case

**Comment by idank** [comment]
@joepavitt How do you configure posthog with vitepress to begin with? [Posthog docs](https://posthog.com/docs/libraries/vue-js) talk about vue plugins etc., but does vitepress allow this?

**Comment by joepavitt** [comment]
PostHog is just a script injection into our "Node-RED Dashboard" documentation. Think PostHog like a Google Analytics.

It's the documentation of Node-RED Dashboard that uses Vitepress

**Comment by idank** [comment]
I see, so you're essentially some code into each page.

I don't know if this would make a difference, but have you tried initializing it per the instructions, through Vue apis? Not sure Vitepress gives access to these though, hence my question.

**Comment by joepavitt** [comment]
> I see, so you're essentially injecting some code into each page.

Yep, just into the `<head>` of the pages.

> I don't know if this would make a difference, but have you tried initializing it per the instructions, through Vue apis? Not sure Vitepress gives access to these though, hence my question.

I have not, although with PostHog, we use it across many applications and websites, in none of these do we have this problem. Only when working with vitepress, which I narrowed down to the this router problem/cause.
